APE26: Removing data storage (representations) from coordinate frames (#112)
astropy/astropy-APEs: APE26.rst

Removing data storage (representations) from coordinate frames
==============================================================

Authors (alphabetical): Jeff Jennings, Adrian Price-Whelan, Nathaniel Starkman, Marten van Kerkwijk
---------------------------------------------------------------------------------------------------

:date-created: 2024 11 04
:date-last-revised: 2024 11 04
:date-accepted: 202x xx xx
:type: Standard Track
:status: Discussion

> Abstract

Author: You're right - I've added text to give the context of APE 5, and to motivate what we propose changing and why, in [46a6d82](https://github.com/astropy/astropy-APEs/pull/112/commits/46a6d822a53863f38951b848e21e9a94cb111213).

> Abstract

Reviewer: Should APE 5 also be updated slightly to link to this one in this PR, so that in the future, someone who stumbles upon APE 5 knows to also go here for the rest of the story?

functionality with slightly different APIs, and better satisfy the principle of 
Separation of Concerns.

Detailed description
--------------------
The coordinate frame classes (subclasses of ``BaseCoordinateFrame``, e.g., ``ICRS``, 
``Galactic``, ``FK4``, etc.) are used to represent astronomical reference frames. These 
may also contain position, velocity, time, or other information about the reference frame 
relative to another reference frame, which is used to transform coordinates between the 
reference frames that exist in the ``astropy.coordinates`` ecosystem. For example, the 
``AltAz`` frame class, which is used when referencing sky coordinates in altitude and 
azimuth, must contain a location on Earth and a time in order to transform to and from a 
fixed reference frame like the ICRS. In addition to reference frame information, these 
frame classes can **also** contain coordinate data (positions, velocities, or other 
differentials), which are stored internally using the ``Representation`` and 
``Differential`` classes.

The above reveals a problematic oddity with the coordinate frame implementation – an 

> insufficient `separation of concerns <https://en.wikipedia.org/wiki/Separation_of_concerns>`_. 

Author: That's a good point - I've reworded the text to reflect this in [46a6d82](https://github.com/astropy/astropy-APEs/pull/112/commits/46a6d822a53863f38951b848e21e9a94cb111213).

``Quantity`` is like ``SkyCoord``.

Having both the frame classes as well as ``SkyCoord`` be able to store and handle data 
has resulted in a large amount of code duplication. Restructuring the ``Frame`` classes 
to remove data storage will allow for much more maintainable, de-duplicated code. It 
will also make it easier to contribute: if there is a problem one would like to solve 
in a given method, if one looks in ``SkyCoord``, one will likely find that it does not 
exist, but instead is defined on ``BaseCoordinateFrame`` and gets dynamically called via 
``SkyCoord.__getattr__``. Indeed, the construction of ``BaseCoordinateFrame`` ends up 
complicating ``SkyCoord``, which has to manage the coordinate data **through** the stored 
``BaseCoordinateFrame``.

Another issue with the current implementation of coordinate frames is that the optional 
inclusion of coordinate data makes the reference frames “multi-modal”. This creates 
different usage modes (with and without data), each exhibiting different behavior. For 
instance, some methods such as ``separation`` work fine with frames with data, while 
doing this on coordinate frames without data results in an error. This kind of 

> multi-modal structure is considered an anti-pattern because it forces any code 

Author: Thanks Erik - I've reworked the text to reflect this in [46a6d82](https://github.com/astropy/astropy-APEs/pull/112/commits/46a6d822a53863f38951b848e21e9a94cb111213) (including explicitly noting the design choice we propose that departs from APE 5, as you suggested).

interacting with the frame classes to handle both cases (checking ``frame.has_data``), 
complicating the codebase unnecessarily.

All the points discussed thus far – separation of concerns and code duplication – 
concern maintainers. However user experience is the more important consideration. In 
this arena too, separating frames from data storage has its advantages. Perhaps most 
importantly, documentation will be more obvious: the methods and attributes are defined 
on ``SkyCoord``, and sphinx will know how to typeset those. It will also be easier: 
following the Zen of Python, there should be one clear way to do something. The present 
overlap leads to confusion where beginner users end up creating ``BaseCoordinateFrame`` 

> objects such as ``ICRS``, when the docs are clear that these are for more advanced users 

Author: I can see how this is a feature, but I think the improved user clarity is motivated earlier on in the text - currently, a user can initialize an `ICRS` object or a `SkyCoord` object, which "represent the same thing, but return different objects with similar but not identical APIs". The proposed change gives only way to have an object hold both data and frame information - `SkyCoord` - preventing user confusion about subtle differences between `SkyCoord` and a with-data frame like the current `ICRS` (including which one to use, and why their API differs in perhaps unexpected ways).

> objects such as ``ICRS``, when the docs are clear that these are for more advanced users 

Reviewer 2: Ah, I see that... But I'm a bit concerned that this still doesn't address the fact that there was a *reason* for this (i.e., they are intentionally different APIs because the objects behave differently when transformed)

- `Allow BaseCoordinateFrames to be stored in tables (by giving them .info) #16831 <https://github.com/astropy/astropy/pull/16831>`_
- `Masked frames and SkyCoord #17106 <https://github.com/astropy/astropy/pull/17016>`_ (this was later removed and instead methods were duplicated)

Further, pull requests have added methods to make frames and ``SkyCoord`` even more 
similar, underscoring that frames *with* data should not be separate entities from 
``SkyCoord``:

- `Implement BaseCoordinateFrame.to_table() #17009 <https://github.com/astropy/astropy/pull/17009>`_
- `Implement BaseCoordinateFrame.frame property #16356 <https://github.com/astropy/astropy/pull/16356>`_

Implementation
--------------
The direct use of coordinate frames instead of ``SkyCoord`` is common. In particular 
``ICRS`` objects are frequently created with data. Given the prevalent use, it is imperative 
to maintain backward compatibility and not break the API too quickly. Therefore, we 
propose implementing this APE through 3 steps (and substeps).

1. Splitting the frame classes into two hierarchies: ones with and without data.

2. Switching ``SkyCoord`` to use the data-less frame classes, and enabling automatic 

> conversion of the with-data frames into ``SkyCoord`` objects.

Author: There wouldn't be duplication because the coordination conversion machinery would be in the new (data-less) frame classes, and the legacy (with-data) frame classes would subclass the new (data-less) classes.

modularly structured, where the coordinate frame class **only** defines the reference 
frame. It would then be the purview of another class – `SkyCoord` – to bring these 
concerns together and to represent data (using `Representation` and `Differential` 
classes) in a given reference frame (using a `CoordinateFrame` class). As a demonstration 
of the current state of duplicated functionality, these two initializations effectively 
represent the same thing, but return different objects with similar but not identical 
APIs:

.. code-block:: python

    c1 = SkyCoord(1., 2., frame="icrs", units="deg")
    c2 = ICRS(1. * u.deg, 2. * u.deg)

We suggest the situation should instead be analogous to what is the case for units, 
which know how to transform from one to another, but are not concerned with how the 
values are stored (that belongs to ``Quantity``). Translating to coordinates, units are 
like ``CoordinateFrame``, the values are the coordinate data (``Representations``), and 
``Quantity`` is like ``SkyCoord``.

Having both the frame classes as well as ``SkyCoord`` be able to store and handle data 

> has resulted in a large amount of code duplication. Restructuring the ``Frame`` classes 

Reviewer 3: Not just code. It's even worse in tests because we have to test both `SkyCoord` and `BaseCoordinateFrame` methods with both `SkyCoord` and `BaseCoordinateFrame` inputs. Not testing all the combinations has led to bugs going unnoticed, e.g. https://github.com/astropy/astropy/pull/15659

Author: Thanks Eero. Marten has amended the text to add this in [9e33a95](https://github.com/astropy/astropy-APEs/pull/112/commits/9e33a95f4e74054a022dbbfc09788343bef219b8) - so I'm going to mark this as resolved.

``SkyCoord.__getattr__``. Indeed, the construction of ``BaseCoordinateFrame`` ends up 
complicating ``SkyCoord``, which has to manage the coordinate data **through** the stored 
``BaseCoordinateFrame``.

Another issue with the current implementation of coordinate frames is that the optional 
inclusion of coordinate data makes the reference frames “multi-modal”. This creates 
different usage modes (with and without data), each exhibiting different behavior. For 
instance, some methods such as ``separation`` work fine with frames with data, while 
doing this on coordinate frames without data results in an error. This kind of 
multi-modal structure is considered an anti-pattern because it forces any code 
interacting with the frame classes to handle both cases (checking ``frame.has_data``), 
complicating the codebase unnecessarily.

All the points discussed thus far – separation of concerns and code duplication – 
concern maintainers. However user experience is the more important consideration. In 
this arena too, separating frames from data storage has its advantages. Perhaps most 
importantly, documentation will be more obvious: the methods and attributes are defined 

> on ``SkyCoord``, and sphinx will know how to typeset those. It will also be easier: 

Reviewer 3: This is relevant not only for documentation but also for type annotations. Looking up `BaseCoordinateFrame` methods via `SkyCoord.__getattr__()` is too dynamic for static type checkers. It is possible to partially alleviate this by writing out stubs in `SkyCoord` (e.g. https://github.com/astropy/astropy/blob/24d48a9e2ef90af5383bf76e6d098e6d193d8186/astropy/coordinates/sky_coordinate.py#L178-L181), but type checkers cannot ensure that the stubs agree with the signatures of the actual methods.

Author: Great point, Marten has amended the text to note this in [9e33a95](https://github.com/astropy/astropy-APEs/pull/112/commits/9e33a95f4e74054a022dbbfc09788343bef219b8) - so I'm going to mark this as resolved.

utility of the proposed changes.

Several issues have been raised regarding topics such as confusion differentiating the 
use of ``frame`` and ``SkyCoord`` for data storage, and problems arising in other astropy 
subpackages when using frames that store data. For example:

- *Comparing Frame with data and SkyCoord with same data raises exception*: `https://github.com/astropy/astropy/issues/13476`
- *Add Frame objects without data to a Table*: `https://github.com/astropy/astropy/issues/16823`

Additionally, multiple pull requests have factored out common code between frames and 
``SkyCoord``, showing that there is no proper separation of concern:

- *Introduction of shared `CoordinateFrameInfo`*: `https://github.com/astropy/astropy/pull/16831`
- *Introduction of `CoordinateSharedMethods`* (this was later removed and instead methods 
were duplicated): `https://github.com/astropy/astropy/pull/17016`

Further, pull requests have added methods to make frames and ``SkyCoord`` even more 
similar, underscoring that frames *with* data should not be separate entities from 
``SkyCoord``:

- *Add .to_table() to frames*: `https://github.com/astropy/astropy/pull/17009`

> - *Add .frame attribute to frames*: `https://github.com/astropy/astropy/pull/16356`

Reviewer 3: I find it confusing that you have not used the pull request titles.

Author: Good point, thank you. Fixed in [46f503e](https://github.com/astropy/astropy-APEs/pull/112/commits/46f503ee97e7ec3f1d050b29b6949a52a48272c2)

the data-less ones getting new names.

2. Switching ``SkyCoord`` to use the data-less frame classes, and enabling automatic
conversion of the with-data frames into ``SkyCoord`` objects.

3. Deprecating the legacy with-data frame classes.

   - Emitting warnings when instantiated.

   - Still warn, but return a ``SkyCoord``, not an instance of its class type (by overriding ``__new__``)

   - Remove.

The third step (at stage 3a) is illustrated in the following pseudocode:

.. code-block:: python

    # === Reference Frame (no data) ===

    class BaseFrame:
        ...

        # Like unit.to(new_unit, value)
        def transform_data_to(self, frame: BaseFrame, data: BaseRepresentation) -> BaseRepresentation:
            """Used by BaseCoordinate for transformation."""
            ...

    class ICRSFrame(BaseFrame):
        pass  # no frame attributes

    class FK5Frame(BaseFrame):
        equinox: TimeAttribute

    # === Coordinates (data + frame) ===

    class BaseCoordinate:
        """Base class for data in a reference frame."""
        frame: BaseFrame
        data: BaseRepresentation
        ...

> 
>     class SkyCoord(BaseCoordinate):

Reviewer 4: ```suggestion

class Coordinate(BaseCoordinate):
    """Data in a reference frame."""
    pass

    class SkyCoord(BaseCoordinate):
         """Data in a reference frame, batteries included."""
```

        def __init__(...):
            # If the frame is a LegacyBaseCoordinateFrame then it is
            # split into a BaseFrame and BaseRepresentation.
            ...

    # === Legacy Coordinate Classes ===

    class BaseCoordinateFrame(BaseCoordinate):

        def __new__(self):
            warnings.warn("Please use SkyCoord")

        @abstractpropery # implemented on subclasses
        def frame(self) -> BaseFrame:
            ...

>     class ICRS(LegacyCoordinateFrame, ICRSFrame):

Reviewer 5: ```suggestion
    class ICRS(BaseCoordinateFrame, ICRSFrame):
```

        ...

>     class FK5(LegacyCoordinateFrame, FK5Frame):

Reviewer 5: ```suggestion
    class FK5(BaseCoordinateFrame, FK5Frame):
```

Removing data storage (representations) from coordinate frames
==============================================================

Authors (alphabetical): Jeff Jennings, Adrian Price-Whelan, Nathaniel Starkman, Marten van Kerkwijk
---------------------------------------------------------------------------------------------------

:date-created: 2024 11 04
:date-last-revised: 2025 04 28
:date-accepted: 202x xx xx
:type: Standard Track
:status: Discussion

Abstract
--------
Following the rationale presented in `APE 5 
<https://github.com/astropy/astropy-APEs/blob/main/APE5.rst>`_, 
coordinate frames in astropy.coordinates currently store metadata 
used to construct the frame (i.e. for transforming between frames) 
and may also store coordinate *data* itself. This duplicates 
functionality with ``SkyCoord``, which acts as a container for both 
coordinate data and reference frame information. We propose to 
change the frame classes such that they only store metadata and 
never coordinate data, superseding this aspect of the implementation 
in `APE 5 <https://github.com/astropy/astropy-APEs/blob/main/APE5.rst>`_. 

> This would make the implementation more modular and performant, 

Reviewer 4: > The other reasons you are listing here should be convincing enough.

Agreed, but also...

>>> you can back this claim with measurements

But we can know this even without. Number of times an object is constructed, going through non-fast-path parsing/validation routes, slotting (easy with dataclases), etc.

> This would make the implementation more modular and performant, 

Reviewer 3: Either back up your claims with actual measurements or don't bring up performance at all. I wouldn't be surprised if the new API allows for faster implementation, but it is not at all obvious to me how much faster it would be. I also suspect that it would be possible to improve performance with the current API (although I don't expect it to be easy because the current API is over-complicated), in which case the performance argument would be weaker.

> This would make the implementation more modular and performant, 

Reviewer 6: Is the plan to only try an implementation for this once the APE is accepted? For similar APEs/PEPs/etc there is sometimes a prototype implementation, which would allow this to be quantified.

> This would make the implementation more modular and performant, 

Reviewer 4: @jeffjennings made https://github.com/jeffjennings/ape26_scratch/blob/main/demo.ipynb :)

> This would make the implementation more modular and performant, 

Reviewer 3: The only measurements in that notebook are in code cell 11 and they are comparing the new `Coordinate` and the current `SkyCoord`. But if `Coordinate` is supposed to be the replacement for the current `BaseCoordinateFrame` instances then that is what you should be comparing `Coordinate` with.

It's fine to mention performance in discussions but I wouldn't bring it up in the APE text itself because performance depends a lot on implementation details that are beyond the scope of an APE.

The coordinate frame classes (subclasses of ``BaseCoordinateFrame``, 
e.g., ``ICRS``, ``Galactic``, ``FK4``, etc.) are used to represent 
astronomical reference frames. These may also contain position, 
velocity, time, or other information about the reference frame 
relative to another reference frame, which is used to transform 
coordinates between the reference frames that exist in the 
``astropy.coordinates`` ecosystem. For example, the ``AltAz`` frame 
class, which is used when referencing sky coordinates in altitude 
and azimuth, must contain a location on Earth and a time in order 
to transform to and from a fixed reference frame like the ICRS. In 
addition to reference frame information, these frame classes can 
**also** contain coordinate data (positions, velocities, or other 
differentials), which are stored internally using the 
``Representation`` and ``Differential`` classes.

The above reflects a design choice in the coordinate frame 
implementation in `APE 5 
<https://github.com/astropy/astropy-APEs/blob/main/APE5.rst>`_ 
that, following several years of use, we have learned from and 
think best to revise. We now view it as an insufficient 
`separation of concerns 

> <https://en.wikipedia.org/wiki/Separation_of_concerns>`_. The 

Reviewer: This link should be moved to where it is first mentioned in the Abstract?

Author: Done in [d801d8d](https://github.com/astropy/astropy-APEs/pull/112/commits/d801d8dba8f52d81878d3b54c384f9a1be8b862b).

frame. It would then be the purview of another class – like 
``SkyCoord`` or the newly-proposed ``Coordinate`` – to bring these 
concerns together and to represent data (using ``Representation`` 
and ``Differential`` classes) in a given reference frame (using a 
``CoordinateFrame`` class). As a demonstration of the current state 
of duplicated functionality, these two initializations effectively 
represent the same thing, but return different objects with similar 
but not identical APIs:

.. code-block:: python

    c1 = SkyCoord(1., 2., frame="icrs", units="deg")
    c2 = ICRS(1. * u.deg, 2. * u.deg)

We suggest the situation should instead be analogous to what is the 
case for units, which know how to transform from one to another, but 
are not concerned with how the values are stored (that belongs to 
``Quantity``). Translating to coordinates, units are like 
``CoordinateFrame``, the values are the coordinate data 
(``Representations``), and ``Quantity`` is like ``SkyCoord``.

Having both the frame classes as well as ``SkyCoord`` be able to
store and handle data has resulted in a large amount of code
duplication (not just in the code proper, but also in the tests,

> leading to missed combinations). Restructuring the ``Frame``

Reviewer 3: The statement in the parenthesis is the opposite of what it should be. Having duplicated test code does not lead to missing coverage. Not having duplicated tests is what has caused problems to go unnoticed.

It is also worth remarking that having duplicated code sometimes means quadrupled (not duplicated) tests because we have to test both `BaseCoordinateFrame` and `SkyCoord` methods with both `BaseCoordinateFrame` and `SkyCoord` arguments.

Reviewer 4: > Not having duplicated tests is what has caused problems to go unnoticed.

👍 😆.

Author: Updated to be correct and add your point in [9332ae6](https://github.com/astropy/astropy-APEs/pull/112/commits/9332ae62c91c54ed09c4c0b64f7c72c6833687cd).

classes to remove data storage will allow for much more
maintainable, de-duplicated code. It will also make it easier to
contribute: if there is a problem one would like to solve in a
given method, if one looks in ``SkyCoord``, one will likely find
that it does not exist, but instead is defined on
``BaseCoordinateFrame`` and gets dynamically called via

> ``SkyCoord.__getattr__``. Indeed, the construction of

Reviewer 3: I have some doubts whether or not potential new contributors manage to find `SkyCoord.__getattr__()`.

Author: Changed to Nathaniel's text in [cd5471b](https://github.com/astropy/astropy-APEs/pull/112/commits/cd5471b1092a3aae4ae31ce9f3b5348df7e91319).

fine with frames with data, while doing this on coordinate frames
without data results in an error. While this multi-modal
structure as motivated in `APE 5
<https://github.com/astropy/astropy-APEs/blob/main/APE5.rst>`_
can be seen as a benefit for interpretability of the logic, we
also now see it from the perspective of an anti-pattern. It
forces any code interacting with the frame classes to handle both
cases (checking ``frame.has_data``), complicating the codebase.
Moreover static analyzers cannot determine which case is being
used; with separated frames and data static analyzers will be
able to prevent this entire class of errors. We thus find it
worthwhile to separate coordinate data from reference frames at
the possible expense of developers having to learn this new
framework.

All the points discussed thus far – separation of concerns and
code duplication – concern maintainers. However user experience
is the more important consideration. In this arena too,
separating frames from data storage has its advantages. Perhaps
most importantly, documentation will be more obvious: the methods
and attributes are defined on ``SkyCoord`` (and ``Coordinate``)

> proper, so sphinx will know how to typeset those, while type

Reviewer: Not very sure what this means but also maybe unnecessary to call out Sphinx specifically, as our code design should not be dictated by our choice of documentation renderer. But if you choose to keep the text, should capitalize and link, as follows.

```suggestion
proper, so `Sphinx <https://www.sphinx-doc.org/>`_ will know how to typeset those, while type
```

checkers can help users in finding and using them properly. It

> will also be easier: following the Zen of Python, there should be

Reviewer: ```suggestion
will also be easier: following the `Zen of Python <https://peps.python.org/pep-0020/>`_, "There should be
```

> one clear way to do something. The present overlap leads to

Reviewer: ```suggestion
one-- and preferably only one --obvious way to do it." The present overlap leads to
```

confusion wherein beginner users end up creating

> ``BaseCoordinateFrame`` objects such as ``ICRS``, when the docs

Reviewer 3: ```suggestion
``BaseCoordinateFrame`` instances, when the docs
```

are clear that these are for more advanced users and that
``SkyCoord`` is to be preferred. The system will also be less
fragile. For example, if users manipulate the internal workings
of ``SkyCoord`` (which is discouraged but possible), the
coordinate data can become decoupled from the caching that

> ``SkyCoord`` performs for speed. With these proposed changes

Reviewer 3: I am not worried about users who mess with `SkyCoord` internals. The problems they will cause will be their own fault. But it might be worth considering downstream developers who subclass `SkyCoord`.

Reviewer 4: That is a better example to highlight.

Author: Updated the text accordingly in [062755f](https://github.com/astropy/astropy-APEs/pull/112/commits/062755fab13a96d06e511507412c04eee929b32e).

users will have a more clear, introspectable, and robust system.

Finished Product
----------------
The end result of the implementation of this APE are be two separate
hierarchies of classes: reference frame classes and coordinate
classes which bring together a reference frame and coordinate
data. We discuss each class type in turn.

Coordinate frame classes only hold information pertaining to
the reference frame they represent and never actual coordinate
data in that reference frame. This is consistent with our
mathematical framework, the reference frame mediates how

> coordinate data is understood (e.g. distance measures) or

Reviewer: ```suggestion
coordinate data is understood (e.g., distance measures) or
```

> interacts (e.g. separation from other coordinates), but the

Reviewer: ```suggestion
interacts (e.g., separation from other coordinates), but the
```

coordinate data itself is actually independent of that
information.

Classes like ``SkyCoord`` will be composed structures bringing
together the reference frame (an instance of a ``BaseFrame``
subclass) and the coordinate data (``BaseRepresentation``
objects). We also introduce a new class, ``Coordinate``, which
is akin to ``SkyCoord`` (containing both frame and data), but
without extra features like keeping frame attributes not
associated with the current frame, caching and flexible input
parsing. In this way ``Coordinate`` operates very similarly to
the current ``BaseCoordinateFrame`` objects when they have
data, and is meant to be their direct replacement in the new

> framework as well as a more lightweight and performant

Reviewer 3: How does the new `Coordinate` class avoid code duplication and user confusion that this APE tries to solve? Why can't we just use `SkyCoord`?

Reviewer 4: > avoid code duplication and user confusion that this APE tries to solve?

Common base class.

Reviewer 3: Am I supposed to learn that from reading this paragraph?

Reviewer 4: Yes, in part. See the lower pseudo-code that demonstrates how simple `Coordinate` will be as a subclass of the common base class. That and the assumption that for changes large enough to merit an APE there would need to be a reasonable amount of explanation. For example, even a `See Also` is very useful for helping users decide when to use a `Coordinate` versus a `SkyCoord`. Combined with further docstring and RTD changes which are the natural course of a PR, and I'm not so worried.

alternative to ``SkyCoord``.

We illustrate the new framework with the following pseudocode.

.. code-block:: python

    class BaseFrame:
        ...

    class ICRSFrame(BaseFrame):
        pass  # no frame attributes

    class FK5Frame(BaseFrame):
        equinox: TimeAttribute

    # ------

    class BaseCoordinate:
        frame: BaseFrame
        data: BaseRepresentation

    class Coordinate(BaseCoordinate):
        ...  # it's fast.

    class SkyCoord(BaseCoordinate):

        def __init__(...):  # flexible input parsing
            ...

Branches and pull requests
--------------------------

> No direct progress on these changes has yet occurred. Discussion

Reviewer 6: How much effort would it be to do a prototype implementation? Are the proposed changes here actually going to amount to a lot of lines of code if we exclude changes to tests, docs, and deprecation warnings/layers?

Author: The prototype has now been implemented (see comment below). The proposed changes here would remove many more lines of legacy code than new lines added (the new `Coordinate` class for example is by design minimal), and the bulk of the refactoring is in moving data-aware methods/attributes/properties out of `SkyCoord` and into a new base class (`BaseCoordinate`).

Reviewer 4: `Coordinate` is fast too! 🎉

  removed and instead methods were duplicated)

Further, pull requests have added methods to make frames and 
``SkyCoord`` even more similar, underscoring that frames *with* 
data should not be separate entities from ``SkyCoord``:

- `Implement BaseCoordinateFrame.to_table() #17009 
  <https://github.com/astropy/astropy/pull/17009>`_
- `Implement BaseCoordinateFrame.frame property #16356 
  <https://github.com/astropy/astropy/pull/16356>`_

In addition, many of these ideas have been developed and tested in
parallal in the JAX-oriented library `coordinax
<https://github.com/GalacticDynamics/coordinax>`_. Many of the
developers of that library are also active Astropy developers and
the development effort towards ``coordinax`` informs, tests, and
validates the ideas presented in this APE. In short, it works.

Implementation
--------------
The direct use of coordinate frames instead of ``SkyCoord`` is 
common. In particular ``ICRS`` objects are frequently created 
with data. Given the prevalent use, it is imperative to maintain 
backward compatibility and not break the API too quickly. 
Therefore, we propose implementing this APE through 4 steps (and 
substeps).

> 1. Splitting the frame classes into two hierarchies: ones with 

Reviewer: For these steps, I would like to know more about what is your proposed timeline and exactly how break-y things will get at each step. And what would we do about helping to notify or patch affected downstream packages. So maybe even list each step as a sub-section if needed.

Even nicer is actual example usage code (not the implementation but *usage* of the implementation) and what the expected output would look like, for an example use case that is very common. Doesn't have to be fancy, just have to reflect what most of us use, which is usually something common like `SkyCoord(ra_arr, dec_arr, unit="deg")`.

Reviewer 7: I think it might still be useful to include the example code that @nstarman wrote above.

Author: Thanks Marten, I've added a section for this in [5a14258](https://github.com/astropy/astropy-APEs/pull/112/commits/5a14258f896596b3bd93fd4110b02e587c3ff5b7), and also added links to the prototype and demo notebook.

   and without data, with the data-less ones getting new names.

2. Adding a new ``Coordinate`` class that is similar to 
   ``SkyCoord``, but which does not keep any frame attributes not 
   in the current frame, and does not have extra features like 
   caching and flexible input parsing. It will only accept 
   data-less frame classes.

3. Switching ``SkyCoord`` to use the data-less frame classes, and 
   enabling automatic conversion of the with-data frames into 
   ``SkyCoord`` objects.

> 4. Deprecating the legacy with-data frame classes.

Reviewer: ```suggestion
4. Deprecating the legacy with-data frame classes, and eventually removing them after a deprecation period that adheres to `APE 2 <https://github.com/astropy/astropy-APEs/blob/main/APE2.rst>`_.
```

   - Emitting warnings when instantiated.

   - Still warn, but return a ``Coordinate``, not an instance of 
     its class type (by overriding ``__new__``)

   - Remove.

> The third step (at stage 3a) is illustrated in the following 

Reviewer: What exactly is `stage 3a` again? I don't see any `a.` sub-bullet until Step 3 above.

Reviewer 4: Yeah, the first sub-bullet in Step 3. Except now it's called Step 4. Needs to be updated.

Author: Done in [a7c80f2](https://github.com/astropy/astropy-APEs/pull/112/commits/a7c80f2b3ce36297d0ee3904b718ef0cf71c34a8).

This would make the implementation more modular and performant, 
remove ambiguity for users from having nearly duplicate functionality 
with slightly different APIs, and better satisfy the principle of 
`Separation of Concerns
<https://en.wikipedia.org/wiki/Separation_of_concerns>`_.

Detailed description
--------------------
The coordinate frame classes (subclasses of ``BaseCoordinateFrame``, 
e.g., ``ICRS``, ``Galactic``, ``FK4``, etc.) are used to represent 
astronomical reference frames. These may also contain position, 
velocity, time, or other information about the reference frame 
relative to another reference frame, which is used to transform 
coordinates between the reference frames that exist in the 
``astropy.coordinates`` ecosystem. For example, the ``AltAz`` frame 
class, which is used when referencing sky coordinates in altitude 
and azimuth, must contain a location on Earth and a time in order 
to transform to and from an inertial reference frame like the ICRS. In 
addition to reference frame information, these frame classes can 
**also** contain coordinate data (positions, velocities, or other 
differentials), which are stored internally using the 

> ``Representation`` and ``Differential`` classes.

Reviewer 3: There is nothing in `astropy.coordinates` called `Representation` or `Differential`.
```suggestion
``BaseRepresentation`` and ``BaseDifferential`` subclasses.
```

The above reflects a design choice in the coordinate frame 
implementation in `APE 5 
<https://github.com/astropy/astropy-APEs/blob/main/APE5.rst>`_ 
that, following several years of use, we have learned from and 
think best to revise. We now view it as an insufficient 
separation of concerns. The 
coordinate frame class deals with two separate issues: the 
definition of the frame and the storage of coordinate data within 
that frame. Ideally the code would be more modularly structured, 
where the coordinate frame class **only** defines the reference 
frame. It would then be the purview of another class – like 
``SkyCoord`` or the newly-proposed ``Coordinate`` – to bring these 
concerns together and to represent data (using ``Representation`` 

> and ``Differential`` classes) in a given reference frame (using a 

Reviewer 3: ```suggestion
concerns together and to represent data (using ``BaseRepresentation`` 
and ``BaseDifferential`` subclasses) in a given reference frame (using a 
```

``CoordinateFrame`` class). As a demonstration of the current state 
of duplicated functionality, these two initializations effectively 
represent the same thing, but return different objects with similar 

> but not identical APIs:

Reviewer 3: Provide examples of the differences.

Author: Added example in [fa2df27](https://github.com/astropy/astropy-APEs/pull/112/commits/fa2df27a8b7001825e1d883be53146ae3ee47ec4)

.. code-block:: python

    c1 = SkyCoord(1., 2., frame="icrs", units="deg")
    c2 = ICRS(1. * u.deg, 2. * u.deg)

We suggest the situation should instead be analogous to what is the 
case for units, which know how to transform from one to another, but 
are not concerned with how the values are stored (that belongs to 
``Quantity``). Translating to coordinates, units are like 
``CoordinateFrame``, the values are the coordinate data 

> (``Representations``), and ``Quantity`` is like ``SkyCoord``.

Reviewer 3: ```suggestion
(``BaseRepresentation``), and ``Quantity`` is like ``SkyCoord``.
```

Having both the frame classes as well as ``SkyCoord`` be able to
store and handle data has resulted in a large amount of code

> duplication. It has also required duplicated or even quadrupled 

Reviewer 3: I don't think code duplication is as big of a problem as it used to be. But reducing code duplication has come at the cost of making `SkyCoord.__getattr__()` even more prevalent. It is worth mentioning that the reliance on `SkyCoord.__getattr__()` means there is at least one method that `SkyCoord` implements accidentally (https://github.com/astropy/astropy/issues/15643) and there might be more.

Author: Added note on this in [8f7711a](https://github.com/astropy/astropy-APEs/pull/112/commits/8f7711a69e41fc14611b0780332f0aaef9b94e24)

tests, in order to test both ``BaseCoordinateFrame`` and 
``SkyCoord`` methods with both ``BaseCoordinateFrame`` and 

> ``SkyCoord`` arguments - where the tests have not caught every 

Reviewer 3: ```suggestion
``SkyCoord`` arguments - where the tests have not covered every 
```

contribute: if there is a problem one would like to solve in a
given method, if one looks in ``SkyCoord``, one will likely find
that it does not exist, and might struggle to find that instead 
it is defined on ``BaseCoordinateFrame`` and gets dynamically 
called via  ``SkyCoord.__getattr__``. Indeed, the construction of
``BaseCoordinateFrame`` ends up complicating ``SkyCoord``, which
has to manage the coordinate data **through** the stored
``BaseCoordinateFrame``.

Another issue with the current implementation of coordinate
frames is that the optional inclusion of coordinate data makes
the reference frames “multi-modal”. This creates different usage
modes (with and without data), each exhibiting different
behavior. For instance, some methods such as ``separation`` work
fine with frames with data, while doing this on coordinate frames
without data results in an error. While this multi-modal
structure as motivated in `APE 5
<https://github.com/astropy/astropy-APEs/blob/main/APE5.rst>`_
can be seen as a benefit for interpretability of the logic, we
also now see it from the perspective of an anti-pattern. It
forces any code interacting with the frame classes to handle both

> cases (checking ``frame.has_data``), complicating the codebase.

Reviewer 3: I suspect that in practice it is more common to avoid code complications by simply assuming that all frames have data, or that none of the frames have data.

Author: Yes although both cases still have to be supported in the current framework, while the proposed changes would avoid the issue both in practice and conceptually.

hierarchies of classes: reference frame classes and coordinate
classes which bring together a reference frame and coordinate
data. We discuss each class type in turn.

Coordinate frame classes only hold information pertaining to
the reference frame they represent and never actual coordinate
data in that reference frame. This is consistent with our
mathematical framework, the reference frame mediates how
coordinate data is understood (e.g., distance measures) or
interacts (e.g., separation from other coordinates), but the
coordinate data itself is actually independent of that
information.

Classes like ``SkyCoord`` will be composed structures bringing
together the reference frame (an instance of a ``BaseFrame``
subclass) and the coordinate data (``BaseRepresentation``
objects). We also introduce a new class, ``Coordinate``, which
is akin to ``SkyCoord`` (containing both frame and data), but
without extra features like keeping frame attributes not
associated with the current frame, caching and flexible input
parsing. In this way ``Coordinate`` operates very similarly to
the current ``BaseCoordinateFrame`` objects when they have

> data, and is meant to be their direct replacement in the new

Reviewer 3: Is the purpose of `Coordinate` to be a replacement for current frame classes with data? Would it be sensible to define both `SkyCoord` and `Coordinate` if you were designing an API from scratch without any historical baggage?

Reviewer 4: I think so, they aim for slightly different things: a WYSIWYG implementation and something with a bit more user-ergonomics optimizations: more flexible init, caching, etc.
Perhaps the flexible init could be a function, e.g. `make_coordinate()`, but in combination with other features, like caching, IMO a class is a good choice too.

framework as well as a more lightweight and performant
alternative to ``SkyCoord``.

We illustrate the new framework with the following pseudocode.

.. code-block:: python

    class BaseFrame:
        ...

    class ICRSFrame(BaseFrame):
        pass  # no frame attributes

    class FK5Frame(BaseFrame):
        equinox: TimeAttribute

    # ------

    class BaseCoordinate:
        frame: BaseFrame

>         data: BaseRepresentation

Reviewer 3: The pseudocode would be much more informative with the [PEP 695 Type Parameter Syntax](https://peps.python.org/pep-0695/).

Author: Hmm I think the pseudocode is clear in conveying inheritance, but the accompanying prototype now linked in the APE is yet another reference.

            ...

Branches and pull requests
--------------------------
No direct progress on these changes has yet occurred. Discussion
of these ideas has however arisen in multiple issues and pull
requests, demonstrating the need for and utility of the proposed
changes.

Several issues have been raised regarding topics such as
confusion differentiating the use of ``frame`` and ``SkyCoord``
for data storage, and problems arising in other astropy
subpackages when using frames that store data. For example:

- `Comparing Frame with data and SkyCoord with same data raises 
  exception #13476 
  <https://github.com/astropy/astropy/issues/13476>`_
- `Add Frame objects without data to a Table #16823 
  <https://github.com/astropy/astropy/issues/16823>`_

Additionally, multiple pull requests have factored out common 
code between frames and ``SkyCoord``, showing that there is no 
proper separation of concern:

- `Allow BaseCoordinateFrames to be stored in tables (by giving 
  them .info) #16831 <https://github.com/astropy/astropy/pull/16831>`_
- `Masked frames and SkyCoord #17106 
  <https://github.com/astropy/astropy/pull/17016>`_ (this was later 

>   removed and instead methods were duplicated)

Reviewer 3: Use an example that was merged. There should be plenty to choose from.

Author: Sorry if there's any confusion in the text; both examples given ([#16831](https://github.com/astropy/astropy/pull/16831) and [#17016](https://github.com/astropy/astropy/pull/17016)) were merged.

common. In particular ``ICRS`` objects are frequently created 
with data. Given the prevalent use, it is imperative to maintain 
backward compatibility and not break the API too quickly. 
Therefore, we propose implementing this APE through 4 steps (and 
substeps).

1. Splitting the frame classes into two hierarchies: ones with 
   and without data, with the data-less ones getting new names.

2. Adding a new ``Coordinate`` class that is similar to 
   ``SkyCoord``, but which does not keep any frame attributes not 
   in the current frame, and does not have extra features like 
   caching and flexible input parsing. It will only accept 
   data-less frame classes.

3. Switching ``SkyCoord`` to use the data-less frame classes, and 
   enabling automatic conversion of the with-data frames into 
   ``SkyCoord`` objects.

4. Deprecating the legacy with-data frame classes, and eventually 
   removing them after a deprecation period that adheres to 
   `APE 2 <https://github.com/astropy/astropy-APEs/blob/main/APE2.rst>`_.

   - Emitting warnings when instantiated.

   - Still warn, but return a ``Coordinate``, not an instance of 

>      its class type (by overriding ``__new__``)

Reviewer 3: I don't see how overriding `__new__()` in such a manner could possibly be a good idea. If you want to make the change more gradual than APE 2 requires then just make the deprecation period longer.

Author: Added note on this in [ae93078](https://github.com/astropy/astropy-APEs/pull/112/commits/ae930787c8e44db11a01fad1042705863146e4f9)

`generics <https://typing.python.org/en/latest/reference/generics.html>`_.
With the proposed framework in this APE, we no longer find such a 
difficulty; the coordinate frame class only defines the 
reference frame, and it becomes the purview of another class – like 
``SkyCoord`` or the newly-proposed ``Coordinate`` – to bring these 
concerns together and to represent data (using ``BaseRepresentation`` 
and ``BaseDifferential`` subclasses) in a given reference frame (using a 
``CoordinateFrame`` class). As a demonstration of the current state of 
duplicated functionality, these two initializations effectively 
represent the same thing, but return different objects with similar 
but not identical APIs:

.. code-block:: python

    c1 = SkyCoord(1., 2., frame="icrs", units="deg")
    c2 = ICRS(1. * u.deg, 2. * u.deg)

Consequently, as one simple example, user confusion can arise from an 
operation as common as transforming reference frames:

.. code-block:: python     

    c1g = c1.transform_to(Galactic())  # Works
    c2g = c2.transform_to(Galactic())  # Works
    c1g = c1.transform_to("galactic")  # Works
    c2g = c2.transform_to("galactic")  # Raises ConvertError

> We suggest the situation should instead be analogous to what is the 

Reviewer 8: This sentence is a little awkward.  Maybe instead: "We suggest the situation should instead be analogous to the structure implemented for astropy.units:  The ``Unit`` class is able to transform between units, but it is not concerned with how the associated values are stored (which is instead handled by ``Quantity``)."

Author: Great, thanks Kyle! Updated in [a6fa590](https://github.com/astropy/astropy-APEs/pull/112/commits/a6fa590b32312305f1f792e25eed31c027a02e92)

tests, in order to test both ``BaseCoordinateFrame`` and 
``SkyCoord`` methods with both ``BaseCoordinateFrame`` and 
``SkyCoord`` arguments - where the tests have not covered every 
combination, problems have gone unnoticed. Restructuring the 
``frame`` classes to remove data storage will allow for much more
maintainable, de-duplicated code. It will also make it easier to
contribute: if there is a problem one would like to solve in a
given method, if one looks in ``SkyCoord``, one will likely find
that it does not exist, and might struggle to find that instead 
it is defined on ``BaseCoordinateFrame`` and gets dynamically 
called via  ``SkyCoord.__getattr__``. Indeed, the construction of
``BaseCoordinateFrame`` ends up complicating ``SkyCoord``, which
has to manage the coordinate data **through** the stored
``BaseCoordinateFrame``.

Another issue with the current implementation of coordinate
frames is that the optional inclusion of coordinate data makes
the reference frames “multi-modal”. This creates different usage
modes (with and without data), each exhibiting different
behavior. For instance, some methods such as ``separation`` work

> fine with frames with data, while doing this on coordinate frames

Reviewer 8: Consider something like: "work fine for coordinate-frame instances that contain data, but they lead to faults for instances without data."

Author: Updated in [a6fa590](https://github.com/astropy/astropy-APEs/pull/112/commits/a6fa590b32312305f1f792e25eed31c027a02e92)

most importantly, documentation will be more obvious: the methods
and attributes are defined on ``SkyCoord`` (and ``Coordinate``)
proper, so `Sphinx <https://www.sphinx-doc.org/>`_ will know how 
to typeset those, while type
checkers can help users in finding and using them properly. It
will also be easier: following the 
`Zen of Python <https://peps.python.org/pep-0020/>`_, "There should 
be one-- and preferably only one --obvious way to do it." The 
present overlap leads to
confusion wherein beginner users end up creating
``BaseCoordinateFrame`` instances, when the docs
are clear that these are for more advanced users and that
``SkyCoord`` is to be preferred. The system will also be less
fragile; with these proposed changes, users - and importantly 
downstream developers who subclass ``SkyCoord`` - will have a 
more clear, introspectable, and robust system.

Finished Product
----------------
The end result of the implementation of this APE will be two separate
hierarchies of classes: reference frame classes and coordinate
classes which bring together a reference frame and coordinate
data. We discuss each class type in turn.

> Coordinate frame classes only hold information pertaining to

Reviewer 8: Here you use "coordinate frame classes," but you use "reference frame classes" above.  It might help clarity and readability to be more (or perhaps overly) consistent with the nomenclature.

Author: Updated in [a6fa590](https://github.com/astropy/astropy-APEs/pull/112/commits/a6fa590b32312305f1f792e25eed31c027a02e92)

        ...

    class SkyCoord(BaseCoordinate):
         """Data in a reference frame, batteries included."""

        def __init__(...):  # flexible input parsing
            # If the frame is a LegacyBaseCoordinateFrame then it is
            # split into a BaseFrame and BaseRepresentation.
            ...

        _cache: dict[str, Any]  # cache

    class Coordinate(BaseCoordinate):
        """Data in a reference frame."""
        ...  # Direct and fast.

    # === Legacy Coordinate Classes ===

    class BaseCoordinateFrame(BaseCoordinate):
        """Reference frames (with optional data storage)."""

        def __new__(self):
            warnings.warn("Please use SkyCoord")

        @abstractpropery # implemented on subclasses
        def frame(self) -> BaseFrame:
            ...

    class ICRS(BaseCoordinateFrame, ICRSFrame):
        ...

    class FK5(BaseCoordinateFrame, FK5Frame):
        ...

We also provide a fuller `prototype 
<https://github.com/jeffjennings/ape26_scratch/blob/main/ape26_prototype.py>` 
and a
`demo notebook <https://github.com/jeffjennings/ape26_scratch/blob/main/demo.ipynb>`
implementing these changes.

Usage patterns
--------------

> The following pseudocode presents typical use cases of the 

Reviewer 8: I suggest including a reference to this section when you first introduce the implementation steps so that readers know examples are provided.

Author: Updated in [a6fa590](https://github.com/astropy/astropy-APEs/pull/112/commits/a6fa590b32312305f1f792e25eed31c027a02e92)
